CMakeLists.txt

```diff
@@ -0,0 +1,250 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2013-2022 The Foundry Visionmongers Ltd
+
+#-----------------------------------------------------------------------
+# CMake preamble.
+
+# FindPython's Development.Module component was added in 3.18, and
+# pybind11 recommends 3.18.2+.
+cmake_minimum_required(VERSION 3.18.2)
+
+# Additional include directories for CMake utils.
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake)
+
+if (APPLE AND NOT DEFINED CMAKE_OSX_DEPLOYMENT_TARGET)
+    # The value of this variable should be set prior to the first
+    # project() command invocation because it may influence
+    # configuration of the toolchain and flags.
+    set(CMAKE_OSX_DEPLOYMENT_TARGET "10.13" CACHE STRING "Minimum OS X deployment version")
+endif ()
+
+# Forbid in-source builds.
+set(CMAKE_DISABLE_IN_SOURCE_BUILD ON)  # NB: currently undocumented.
+# Forbid CMake from editing sources
+set(CMAKE_DISABLE_SOURCE_CHANGES ON)  # NB: currently undocumented.
+
+# By default, do not output "Up-to-date" messages during install. Can
+# be spammy when this project is included in another via CMake's
+# ExternalProject.
+if (NOT DEFINED CMAKE_INSTALL_MESSAGE)
+    set(CMAKE_INSTALL_MESSAGE LAZY)
+endif ()
+
+
+#-----------------------------------------------------------------------
+# Set a default build type if none was specified
+
+if (NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
+    message(STATUS "Setting build type to 'Release' as none was specified.")
+    set(CMAKE_BUILD_TYPE Release CACHE STRING "Choose the type of build." FORCE)
+    # Set the possible values of build type for cmake-gui, ccmake
+    set_property(
+        CACHE CMAKE_BUILD_TYPE
+        PROPERTY STRINGS
+        "Debug"
+        "Release"
+        "MinSizeRel"
+        "RelWithDebInfo")
+endif ()
+
+
+#-----------------------------------------------------------------------
+# IDE helpers
+
+# Generate compile_commands.json to make it easier to work with clang
+# based tools. Note: `project(...)` invisibly defines
+# `CMAKE_EXPORT_COMPILE_COMMANDS` so this check must come before.
```

Comment:
Didn't spot this on the first pass either.

This https://gitlab.kitware.com/cmake/cmake/-/issues/16588 says that if the variable is set after `project()` then it avoids problems.

Was this code copied from elsewhere?
Was it fixing a known problem?

Reply:
Interesting, I saw the exact opposite behaviour they describe, where setting it after `project()` would cause the value to be ignored.  They do say in that issue that the bug has been fixed, in "Milestone" 3.18.0.  

To double-check, I confirmed that it currently works as expected. I.e. I don't see the bug they describe in the issue. Tested CMake 3.21.1, then downgraded to 3.18.2, which I should be using anyway, and confirmed it still works. CMake 3.18.2 was released 20 Aug 2020 (according to github), which is a bit older than the last few of the comments in that issue, but perhaps they were still using an older version?

Reply:
I will admit to never having used it before. If I need to see compile commands, I'm usually digging elsewhere.

But I do see the benefit when it's needed by clang tools as there are some useful ones there.

Let's leave it in.